File: npmjs/pray-cli/src/resolve/upstream-latest.ts
import { readFileSync } from "node:fs";
import { latestConstraintForPackage, versionSatisfies } from "../constraint.js";
import { PrayError } from "../errors.js";
import { prepareGitSources } from "../git/sources.js";
import type { Lockfile } from "../lockfile/types.js";
import type { ManifestPackage } from "../manifest/types.js";
import { findPrayspecFile } from "../package-spec/discovery.js";
import { parsePackageSpec } from "../package-spec/index.js";
import { renderPackageSpec } from "../package-spec/render.js";
import { writeProjectFile } from "../transaction/index.js";
import type { ResolveOptions } from "./context.js";
import { impliedSourceName } from "./package-root.js";
import { resolvePackage } from "./project.js";
import { sourceMap } from "./source-map.js";
import type { ResolvedPackage, ResolvedProject } from "./types.js";

export interface PathUpstreamLatestConstraint {
  packageName: string;
  upstreamName: string;
  currentConstraint: string;
  latestVersion: string;
  newConstraint: string;
  packageRoot: string;
}

export async function planPathUpstreamLatestConstraints(
  project: ResolvedProject,
  previous: Lockfile | undefined,
  selected: string | undefined,
  options: ResolveOptions,
): Promise<PathUpstreamLatestConstraint[]> {
  const gitSources = prepareGitSources(
    project.projectRoot,
    project.manifest.sources,
    previous,
    options.refreshSourceRevisions,
  );
  const sources = sourceMap(project.manifest.sources);
  const plans: PathUpstreamLatestConstraint[] = [];
  for (const packageEntry of project.packages) {
    if (selected && packageEntry.declaration.name !== selected) {
      continue;
    }
    const plan = await planOnePathUpstream(
      project,
      packageEntry,
      sources,
      gitSources,
      previous,
      options,
    );
    if (plan) {
      plans.push(plan);
    }
  }
  return plans;
}

export function applyPathUpstreamLatestConstraints(
  plans: readonly PathUpstreamLatestConstraint[],
): void {
  for (const plan of plans) {
    const specPath = findPrayspecFile(plan.packageRoot);
    const spec = parsePackageSpec(readFileSync(specPath, "utf8"));
    if (!spec.upstream) {
      throw PrayError.resolution(
        `package ${plan.packageName} has no upstream pin`,
      );
    }
    spec.upstream = {
      ...spec.upstream,
      constraint: plan.newConstraint,
    };
    writeProjectFile(specPath, renderPackageSpec(spec));
  }
}

async function planOnePathUpstream(
  project: ResolvedProject,
  packageEntry: ResolvedPackage,
  sources: Map<string, import("../manifest/types.js").ManifestSource>,
  gitSources: ReturnType<typeof prepareGitSources>,
  lockfile: Lockfile | undefined,
  options: ResolveOptions,
): Promise<PathUpstreamLatestConstraint | undefined> {
  const upstream = packageEntry.spec.upstream;
  if (!upstream || !packageEntry.declaration.path) {
    return undefined;
  }
  const source = impliedSourceName(
    namedPackage(upstream.name, "*", undefined),
    sources,
  );
  const resolved = await resolvePackage(
    project.projectRoot,
    sources,
    gitSources,
    namedPackage(upstream.name, "*", source),
    lockfile,
    options,
  );
  if (versionSatisfies(resolved.spec.version, upstream.constraint)) {
    return undefined;
  }
  const newConstraint = latestSpecUpstreamConstraint(
    upstream.constraint,
    resolved.spec.version,
  );
  if (!versionSatisfies(resolved.spec.version, newConstraint)) {
    throw PrayError.resolution(
      `derived constraint ${newConstraint} does not admit latest ${resolved.spec.version} for ${packageEntry.declaration.name}`,
    );
  }
  return {
    packageName: packageEntry.declaration.name,
    upstreamName: upstream.name,
    currentConstraint: upstream.constraint,
    latestVersion: resolved.spec.version,
    newConstraint,
    packageRoot: packageEntry.root,
  };
}

export function latestSpecUpstreamConstraint(
  current: string,
  latestVersion: string,
): string {
  const derived = latestConstraintForPackage(current, latestVersion);
  if (derived.startsWith("=")) {
    return `= ${latestVersion}`;
  }
  return derived;
}

function namedPackage(
  name: string,
  constraint: string,
  source: string | undefined,
): ManifestPackage {
  return {
    name,
    constraint,
    ...(source ? { source } : {}),
    exports: [],
    targets: [],
    features: [],
    groups: [],
    optional: false,
    roles: [],
    bound: false,
  };
}

File: rubygems/pray-cli/lib/pray/upstream_latest.rb
# frozen_string_literal: true

module Pray
  module Upstream
    module_function

    PathUpstreamLatestConstraint = Struct.new(
      :package_name, :upstream_name, :current_constraint, :latest_version, :new_constraint, :package_root,
      keyword_init: true
    )

    def plan_path_upstream_latest_constraints(project, previous, selected, options)
      user_config = Config.load_user_config
      git_sources = GitSources.prepare_git_sources(
        project.project_root,
        project.manifest.sources,
        previous,
        refresh: options.refresh || options.refresh_source_revisions,
        offline: options.offline
      )
      sources = Resolve.source_map(project.manifest.sources)
      plans = []
      project.packages.each do |package|
        next if selected && package.declaration.name != selected

        plan = plan_one_path_upstream(project, package, sources, git_sources, user_config, previous, options)
        plans << plan if plan
      end
      plans
    end

    def apply_path_upstream_latest_constraints(plans)
      plans.each do |plan|
        spec_path = Resolve.find_prayspec_file(plan.package_root)
        spec = Pray.parse_package_spec(File.read(spec_path))
        unless spec.upstream
          raise Error.resolution("package #{plan.package_name} has no upstream pin")
        end

        spec.upstream = PackageUpstream.new(
          name: spec.upstream.name,
          constraint: plan.new_constraint
        )
        Transaction.write_file(spec_path, PackageSpecRender.render_package_spec(spec))
      end
    end

    def latest_spec_upstream_constraint(current, latest_version)
      derived = Constraint.latest_constraint_for_package(current, latest_version)
      return "= #{latest_version}" if derived.start_with?("=")

      derived
    end

    def plan_one_path_upstream(project, package, sources, git_sources, user_config, lockfile, options)
      upstream = package.spec.upstream
      return unless upstream && package.declaration.path

      source = ResolveSource.implied_source_name(ManifestPackage.new(name: upstream.name), sources)
      resolved = Resolve.resolve_package(
        project.project_root,
        sources,
        git_sources,
        user_config,
        ManifestPackage.new(name: upstream.name, constraint: "*", source: source),
        lockfile,
        options: options
      )
      return if Constraint.version_satisfies(resolved.spec.version, upstream.constraint)

      new_constraint = latest_spec_upstream_constraint(upstream.constraint, resolved.spec.version)
      unless Constraint.version_satisfies(resolved.spec.version, new_constraint)
        raise Error.resolution(
          "derived constraint #{new_constraint} does not admit latest #{resolved.spec.version} for #{package.declaration.name}"
        )
      end

      PathUpstreamLatestConstraint.new(
        package_name: package.declaration.name,
        upstream_name: upstream.name,
        current_constraint: upstream.constraint,
        latest_version: resolved.spec.version,
        new_constraint: new_constraint,
        package_root: package.root
      )
    end
    private_class_method :plan_one_path_upstream
  end
end
